relay/relay.go

```diff
@@ -263,261 +242,180 @@ func (r *Relay) process(parentCtx context.Context) error {
-		eventFormat     *replication.FormatDescriptionEvent // latest FormatDescriptionEvent, used when re-calculate checksum
+		uuid, latestPos = r.meta.Pos()
+		_, latestGTID   = r.meta.GTID()
 	)
 
-	closeGapSyncer := func() {
-		addRelayFlag = false // reset to false when closing the gap syncer
-		if gapReader != nil {
-			gapReader.Close()
-			gapReader = nil
-		}
-		gapSyncStartPos = nil
-		gapSyncEndPos = nil
+	if r.cfg.EnableGTID && latestPos.Compare(minCheckpoint) <= 0 {
+		log.Warnf("[relay] no corresponding position specified for GTID sets %s, skip recovering", latestGTID)
+		return nil
 	}
 
+	// setup a special writer to do the recovering
+	cfg := &writer.FileConfig{
+		RelayDir: r.meta.Dir(),
+		Filename: latestPos.Name,
+	}
+	writer2 := writer.NewFileWriter(cfg, parser2)
+	err := writer2.Start()
+	if err != nil {
+		return errors.Annotatef(err, "start recover writer for UUID %s with config %+v", uuid, cfg)
+	}
 	defer func() {
-		closeGapSyncer()
-		if r.fd != nil {
-			r.fd.Close()
+		err2 := writer2.Close()
+		if err2 != nil {
+			log.Errorf("[relay] close recover writer for UUID %s with config %+v error %v", uuid, cfg, err2)
 		}
 	}()
-
-	go r.doIntervalOps(parentCtx)
-
-	for {
-		if gapReader == nil && gapSyncStartPos != nil && gapSyncEndPos != nil {
-			gapReader = binlogreader.NewTCPReader(r.gapSyncerCfg)
-			err = gapReader.StartSyncByPos(*gapSyncStartPos)
+	log.Infof("[relay] started recover writer for UUID %s with config %+v", uuid, cfg)
+
+	result, err := writer2.Recover()
+	if err == nil {
+		if result.Recovered {
+			log.Warnf("[relay] relay log file recovered from position %s to %s, GTID sets %v to %v",
+				latestPos, result.LatestPos, latestGTID, result.LatestGTIDs)
+			err = r.meta.Save(result.LatestPos, result.LatestGTIDs)
 			if err != nil {
-				return errors.Annotatef(err, "start to fill gap in relay log file from %v", gapSyncStartPos)
-			}
-			log.Infof("[relay] start to fill gap in relay log file from %v", gapSyncStartPos)
-			if r.cfg.EnableGTID {
-				_, currentGtidSet := r.meta.GTID()
-				if currentGtidSet.Equal(r.gSetWhenSwitch) {
-					addRelayFlag = true
-					log.Infof("[relay] binlog events in the gap for file %s already exists in previous sub directory, adding a LOG_EVENT_RELAY_LOG_F flag to them. current GTID set %s", r.fd.Name(), currentGtidSet)
-				} else if currentGtidSet.Contain(r.gSetWhenSwitch) {
-					// only after the first gap filled, currentGtidSet can be the superset of gSetWhenSwitch
-					log.Infof("[relay] binlog events in the gap for file %s not exists in previous sub directory, no need to add a LOG_EVENT_RELAY_LOG_F flag. current GTID set %s, previous sub directory end GTID set %s", r.fd.Name(), currentGtidSet, r.gSetWhenSwitch)
-				} else {
-					return errors.Errorf("currnet GTID set %s is not equal or superset of previous sub directory end GTID set %s", currentGtidSet, r.gSetWhenSwitch)
-				}
+				return errors.Annotate(err, "save meta after recovered")
```

Comment:
add latest position and latest gtids into error?

Reply:
done in https://github.com/pingcap/dm/pull/171/commits/201f96028d56a0aaa1c17373381a391bdbe8e75a.